Make string() function consider the BOM

danielgindi opened the pull request:
So `response.body().string(true)` will return a string without the BOM.

This helps avoid a lot of boilerplate code just for handling BOMs, in every app that uses OkHttp.

JakeWharton commented:
I think #2707 will solve this automatically. All calls to `string()` will be BOM-free once that is implemented.

JakeWharton commented:
A BOM, by its nature and definition, shouldn't be exposed to the consumer in the char world. Only the byte world (aka `source()` or `byteStream()`).

danielgindi commented:
Tell that to Sun... :)

It's been a month, I thought maybe there are complexities that delay the merge, so I've added a simple function to handle this at least for the `string()`.

If the other one is getting merged for sure, feel free to close this. I hope that this will at least help people coming here looking for a temporary solution, it's a copy-paste function...

swankjesse commented:
I think we want this API. I’m not completely convinced we want to do it here and not in Okio, but this is a good start.

Can you confirm that this is 100% original code written by you, and complete our CLA?
https://github.com/square/okhttp/blob/master/.github/CONTRIBUTING.md

danielgindi commented:
Yes, this is completely written by me. Not too complex, I think :)

danielgindi commented:
And done!

JakeWharton commented:
I don't want this API. This is providing the user a choice when there is no actual choice needed. Just make `string()` do the Right Thing™. If you want a BOM-containing string then use `source().readUtf8()`. `charStream()` also should do the Right Thing™.

swankjesse commented:
Yup, I'm convinced. Let's just eat the BOM in string() and charStream(). If you want it, you do your own InputStreamReader.

danielgindi commented:
Yeah, I thought so too, just didn't want to do introduce any breaking changes. Then I realized that if anyone has considered BOM in their code, they must have done so conditionally.

I've modified the code and rebased.

danielgindi commented:
I've just fixed a typo in my commit, but I see that you have a bigger solution prepared here for chartStream as well.

JakeWharton commented:
Just a heads up I submitted #2823 earlier this morning.
